File: Ising.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define L 10 // Lattice size
#define N (L*L*L) // Number of spins
#define J 1.0 // Interaction energy
#define KB 1.0 // Boltzmann constant

int lattice[L][L][L];

void initialize_lattice() {
    for (int x = 0; x < L; x++) {
        for (int y = 0; y < L; y++) {
            for (int z = 0; z < L; z++) {
                lattice[x][y][z] = 1; // Randomly assign spin up (+1) or down (-1)
            }
        }
    }
}

double calculate_energy() {
    double energy = 0.0;
    for (int x = 0; x < L; x++) {
        for (int y = 0; y < L; y++) {
            for (int z = 0; z < L; z++) {
                int spin = lattice[x][y][z];
                int neighbor_sum = lattice[(x+1)%L][y][z] + lattice[(x-1+L)%L][y][z] +
                                   lattice[x][(y+1)%L][z] + lattice[x][(y-1+L)%L][z] +
                                   lattice[x][y][(z+1)%L] + lattice[x][y][(z-1+L)%L];
                energy -= J * spin * neighbor_sum;
            }
        }
    }
    return energy / 2.0; // Each pair counted twice
}

void metropolis_step(double T) {
    for (int i = 0; i < N; i++) {
        int x = rand() % L;
        int y = rand() % L;
        int z = rand() % L;
        int spin = lattice[x][y][z];
        int neighbor_sum = lattice[(x+1)%L][y][z] + lattice[(x-1+L)%L][y][z] +
                           lattice[x][(y+1)%L][z] + lattice[x][(y-1+L)%L][z] +
                           lattice[x][y][(z+1)%L] + lattice[x][y][(z-1+L)%L];
        double dE = 2.0 * J * spin * neighbor_sum;
        if (dE < 0 || (rand() / (double)RAND_MAX) < exp(-dE / (KB * T))) {
            lattice[x][y][z] = -spin;
        }
    }
}

double calculate_magnetization() {
    double magnetization = 0.0;
    for (int x = 0; x < L; x++) {
        for (int y = 0; y < L; y++) {
            for (int z = 0; z < L; z++) {
                magnetization += lattice[x][y][z];
            }
        }
    }
    return magnetization / N;
}

int main() {
    srand(time(NULL));
    initialize_lattice();

    double T_start = 4.0;
    double T_end = 5.0;
    double T_step = 0.01;

    for (double T = T_start; T <= T_end; T += T_step) {
        for (int i = 0; i < 10000; i++) {
            metropolis_step(T);
        }

        double magnetization = abs(calculate_magnetization());
        printf("Temperature: %f, Magnetization: %f\n", T, magnetization);
    }

    return 0;
}

File: ising.py
import numpy as np
import random

# Parameters
L = 10  # Lattice size
J = 1.0  # Interaction strength
k_B = 1.0  # Boltzmann constant
n_steps = 1000  # Number of Monte Carlo steps
T_min, T_max, T_step = 4.0, 5.0, 0.1  # Temperature range

# Initialize the lattice with random spins
def initialize_lattice(L):
    return np.random.choice([-1, 1], size=(L, L, L))

# Calculate the energy of a given spin configuration
def calculate_energy(lattice, L, J):
    energy = 0
    for x in range(L):
        for y in range(L):
            for z in range(L):
                S = lattice[x, y, z]
                neighbors = lattice[(x+1)%L, y, z] + lattice[x, (y+1)%L, z] + lattice[x, y, (z+1)%L] + \
                            lattice[(x-1)%L, y, z] + lattice[x, (y-1)%L, z] + lattice[x, y, (z-1)%L]
                energy += -J * S * neighbors
    return energy / 2  # Each pair counted twice

# Perform a single Monte Carlo step
def monte_carlo_step(lattice, L, J, T):
    for _ in range(L**3):
        x, y, z = random.randint(0, L-1), random.randint(0, L-1), random.randint(0, L-1)
        S = lattice[x, y, z]
        neighbors = lattice[(x+1)%L, y, z] + lattice[x, (y+1)%L, z] + lattice[x, y, (z+1)%L] + \
                    lattice[(x-1)%L, y, z] + lattice[x, (y-1)%L, z] + lattice[x, y, (z-1)%L]
        dE = 2 * J * S * neighbors
        if dE < 0 or random.random() < np.exp(-dE / (k_B * T)):
            lattice[x, y, z] *= -1

# Main simulation loop
def simulate(L, J, T_min, T_max, T_step, n_steps):
    temperatures = np.arange(T_min, T_max, T_step)
    magnetizations = []

    for T in temperatures:
        lattice = initialize_lattice(L)
        for step in range(n_steps):
            monte_carlo_step(lattice, L, J, T)
        magnetization = np.abs(np.sum(lattice)) / (L**3)
        magnetizations.append(magnetization)
        print(f"T = {T:.2f}, Magnetization = {magnetization:.4f}")

    return temperatures, magnetizations

# Run the simulation
temperatures, magnetizations = simulate(L, J, T_min, T_max, T_step, n_steps)

# Plot the results
import matplotlib.pyplot as plt

plt.plot(temperatures, magnetizations, 'o-')
plt.xlabel('Temperature')
plt.ylabel('Magnetization')
plt.title('3D Ising Model')
plt.show()